### PR TITLE
Initial release of the Tivo 1.1 binding for OpenHAB2

### PR DESCRIPTION
**Pull request by DigitalBites**
Initial discussion and idea generation: https://community.openhab.org/t/tivo-1-1-protocol-new-binding-contribution/5572

Everything works except for the auto-detection which DOES work when the ipp add-on is not enabled. I believe there is a bigger issue with mDNS discovery in the base code that will need to be sorted out. Will continue to debug this but the plugin is functional via manually specifying an IP and adding a device.

### CODE REVIEW

File: addons/binding/org.openhab.binding.tivo/ESH-INF/i18n/tivo_xx_XX.properties [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# binding
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/ESH-INF/i18n/tivo_xx_XX.properties]: please remove this file

File: addons/binding/org.openhab.binding.tivo/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -0,0 +1,30 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: TiVo Binding
+Bundle-SymbolicName: org.openhab.binding.tivo;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ClassPath: .
+Import-Package: 
+ javax.jmdns;version="3.4.2",
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/META-INF/MANIFEST.MF]: please remove the version constraint

File: addons/binding/org.openhab.binding.tivo/META-INF/MANIFEST.MF [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ org.eclipse.smarthome.core.thing,
+ org.eclipse.smarthome.core.thing.binding,
+ org.eclipse.smarthome.core.thing.binding.builder,
+ org.eclipse.smarthome.core.thing.link,
+ org.eclipse.smarthome.core.thing.type,
+ org.eclipse.smarthome.core.types,
+ org.eclipse.smarthome.io.transport.mdns.discovery,
+ org.openhab.binding.tivo,
+ org.openhab.binding.tivo.handler,
+ org.openhab.core.library.types,
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/META-INF/MANIFEST.MF]: please remove this import - there must not be a dependency on this package in org.openhab.

File: addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java [hunk 4]
```diff
@@ -0,0 +1,962 @@
...
+
+        if (myTivo != null) {
+            logger.debug("While refreshing thing '{}' - status data FOUND in CURRENT CHECK - '{}'", getThing().getUID(),
+                    myTivo.toString());
+        } else {
+            logger.debug("While refreshing thing '{}' - status data NOT FOUND in CURRENT CHECK!", getThing().getUID());
+        }
+
+        if (myTivo != null && myTivo.isCmdOk()) {
+            // pTivoData = myTivo;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java]: please remove commented code

File: addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java [hunk 5]
```diff
@@ -0,0 +1,962 @@
...
+        updateStatus();
+
+        return tivoStatusData;
+    }
+
+    private void updateStatus() {
+
+        if (tivoStatusData != null) {
+
+            updateState(new ChannelUID(getThing().getUID(), CHANNEL_TIVO_STATUS),
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java]: please replace these calls by
```
updateState(CHANNEL_TIVO_STATUS, new StringType(tivoStatusData.getMsg()))
```

File: addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java [hunk 6]
```diff
@@ -0,0 +1,962 @@
...
+
+        Configuration conf = editConfiguration();
+        cfgIgnoreChannels.add(pChannel);
+        conf.put(CONFIG_IGNORE_CHANNELS, doCfgParseIgnoreChannel(cfgIgnoreChannels));
+        updateConfiguration(conf);
+
+    }
+
+    private boolean getChkIgnoredChannel(int pChannel) {
+
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java]: remove empty line

File: addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java [hunk 7]
```diff
@@ -0,0 +1,962 @@
...
+        if (cfgIgnoreChannels != null && cfgIgnoreChannels.contains(pChannel)) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private int getNextChannel(int pChannel) {
+
+        // if (getChkIgnoredChannel(pChannel)) {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java]: remove commented code

File: addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java [hunk 8]
```diff
@@ -0,0 +1,962 @@
...
+        if (tivoStatusData != null && tivoStatusData.isCmdOk()) {
+            // retry logic is allowed otherwise we only do the logic below once.
+
+            if (tivoStatusData.getChannelNum() > pChannel && pChannel > 1) {
+                // we appear to be changing the channel DOWNWARD so we try to go down a bit further
+                // addIgnoredChannel(chnl);
+                pChannel--;
+                logger.info("TiVo handler for thing '{}' next proposed channel - '{}'", getThing().getUID(), pChannel);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java]: also only do debug, not info level here

File: addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java [hunk 9]
```diff
@@ -0,0 +1,962 @@
...
+            } else if (tivoStatusData.getChannelNum() < pChannel) {
+                // addIgnoredChannel(chnl);
+                pChannel++;
+                logger.info("TiVo handler for thing '{}' next proposed channel + '{}'", getThing().getUID(), pChannel);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java]: dito

File: addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/discovery/TiVoDiscoveryParticipant.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.io.transport.mdns.discovery.MDNSDiscoveryParticipant;
+import org.openhab.binding.tivo.TiVoBindingConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TiVoDiscoveryParticipant implements MDNSDiscoveryParticipant {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/discovery/TiVoDiscoveryParticipant.java]: please add a short JavaDoc

File: addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoServiceProvider.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+ * Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.tivo.internal.service;
+
+class TivoServiceProvider {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoServiceProvider.java]: completely remove this class, this looks like an accident.

**Comment by kaikreuzer** [comment]
Thanks for your work!
I did a rough initial review. Please remove commented code throughout the PR and also empty lines (which seem to be a few to many in the PR).
If I get it right, the binding is rather work-in-progress, so it might be a good candidate to be marked as an "experimental" feature (a new category I am planning to add) until it is stable and mature. Does this sound ok to you?

**Comment by DigitalBites** [comment]
Thanks for the feedback! I am in full support of it being labeled as experimental however I want to pull this pull request and resubmit (or is there a way to update this one so i don't lose the feedback?)  I am about done going through he code and cleaning a bunch of things up and will take the feedback into account as well.

**Comment by kaikreuzer** [comment]
> or is there a way to update this one so i don't lose the feedback?

Yes! You can rename your existing branch, create a new one with the old name and do a "force push" to your fork - this will automatically update the PR with the content of your branch.

**Comment by kaikreuzer** [comment]
@DigitalBites Any update from your end?

**Comment by kaikreuzer** [comment]
@DigitalBites Please don't let this effort die - hope you find time address my comments soon!

**Comment by jahubba** [comment]
I recently came across this pull request.  I cloned @DigitalBites's repo and was able to successfully build/deploy.  It would be great if this was included in the official repo.  Hopefully and ideally @DigitalBites can get to this, if not, I can try to incorporate the changes.

**Comment by AndyXMB** [comment]
@jahubba I've been doing some work on this binding based on the code that @DigitalBites shared.  I've completed most of the code / implement some functionality that was discussed in the original thread.  I think I am nearly there and am just doing some final testing / adding some Java Docs.

**Comment by jahubba** [comment]
@AndyXMB Great to hear.  Let me know if you would like help testing at all.

**Comment by AndyXMB** [comment]
I will… have one bug left in the new channel scanning feature… of course it is intermittent so trying to work out what is causing it is proving to be a bit of a nightmare.

I am probably going to post a file based “addon” copy of the code first to the original discussion thread first (https://community.openhab.org/t/tivo-1-1-protocol-new-binding-contribution/5572) .  Mainly as I only have one TiVo so need someone to test who has more than one device and I need to figure out how to work Git ;)  Andy

**Comment by AndyXMB** [comment]
@jahubba I've just posted a .jar based version of this.  See the original discussion thread  https://community.openhab.org/t/tivo-1-1-protocol-new-binding-contribution/5572.  Next move is to work out how to get this onto Git etc.  Andy.

**Comment by jahubba** [comment]
Awesome, I'll try out this jar and let you know if I run into any issues. Thanks for adding this.

**Comment by martinvw** [comment]
Looking at the forum @AndyXMB took over further development of this PR, please mention this old one when you create a new one for it, that way we know that we can close this one.

@AndyXMB when do you plan to submit your PR? We might still require some time before merging it because we will have some review comments and there are also some other PR's in the queue.

**Comment by AndyXMB** [comment]
Hi @martinvw , I was hoping to get some feedback on testing from other users before I submitting the PR as this is my first attempt at Java / something like this.  Steep learning curve for me, not sure I have used the best techniques on top of the usual challenged of reverse engineering etc.

Volume of downloads has been low and feedback has been non existent, so honestly I got a bit disheartened and moved onto other things.  

In my own testing things have been working perfectly until the last upgrade of the base code.  The binding is now not initialising correctly, so I need to take a look at the code again and work out what has changed.

So bear with me, it may take me a week or two as I'm also mid way through refurbishing the bathroom!  However this time I will submit a PR when the code is functional and let you guys comment.  Andy

**Comment by jahubba** [comment]
Hi @AndyXMB 

Sorry, I meant to report back with my testing earlier and slipped my mind.  I've been using it for months now without any issues.  I've not ran into the issue you had either, but I haven't upgraded openHAB since originally installing the Tivo binding.

Just a note, I cloned your repo and created the binding myself, I tend not to download jars and try to rebuild myself when I can.  Maybe others do the same and perhaps why downloads are low.

I'd be happy to continue testing or help with any integration work.  It would be nice to get this added so we don't need to install jar manually.

Thanks for the work Andy.

Jason

**Comment by martinvw** [comment]
@jahubba thanks for your contact good to hear that you are using it :-)

**Comment by AndyXMB** [comment]
Thanks for the feedback @jahubba, can you update the main conversation thread  https://community.openhab.org/t/tivo-1-1-protocol-new-binding-contribution/5572 with the model(s) of TiVo that you are controlling.  I'm keen to know if we have confirmation that things are OK in the UK (Virgin Media) and USA models as they are subtly different etc.   Thanks in anticipation. A

PS: the problem I mentioned in the last post was cured by the deletion and recreation (by Discovery) of my TiVo item.  Since doing that, it works perfectly again on  2.1.0~20170430040301-1 (Build #904).  So hopefully all I need to do is make sure I have all the Java Docs done and double check the documentation before I can raise the PR.

**Comment by martinvw** [comment]
Closed and take over by @AndyXMB in #2312

**Pull request closed without merging**